lib/cn/kvset.c

```diff
@@ -1914,35 +1914,37 @@ kvset_get_vbsetv(struct kvset *ks, uint *vbsetc)
 }
 
 /**
- * kvset_get_max_key() - Get the largest key in a kvset
+ * kvset_get_min_key() - Get the smallest key in a kvset
  * @ks:   struct kvset handle.
- * @key:  (output) largest key. Null if kvset contains only ptombs.
- * @klen: (output) length of @key. Zero if kvset contains only ptombs.
+ * @min_key:  (output) smallest key
+ * @min_klen: (output) length of @min_key
  */
 void
-kvset_get_max_key(struct kvset *ks, void **key, uint *klen)
+kvset_get_min_key(struct kvset *ks, void **min_key, uint *min_klen)
 {
-    struct kvset_kblk *kb;
-    int                last = ks->ks_st.kst_kblks - 1;
+    struct kvset_kblk *kb = &ks->ks_kblks[0];
 
-    *key = 0;
-    *klen = 0;
+    INVARIANT(min_key && min_klen);
 
-    /* Check if kblock has non-ptombs. If not, use the second to last kblk.
-     * If there is exactly one kblock and it contains only ptombs, return
-     * with (*key) set to 0;
-     */
-    if (ks->ks_kblks[last].kb_wbt_desc.wbd_n_pages == 0) {
-        if (!last)
-            return;
+    *min_key = (void *)kb->kb_koff_min;
+    *min_klen = kb->kb_klen_min;
+}
 
-        --last;
-    }
+/**
```

Comment:
Put documentation in header file

Reply:
Documentation should mention that key ptr is only valid as long as kvset exists.  Callers must copy it, or keep a ref to the kvset.

Change type of min_key param to reflect that it points to const data.

Reply:
if you have a ref on the kvset you don't need the tree lock, what are the use cases for these functions?

Reply:
ah, you would need refs on all the kvsets which is unweildy...

Reply:
> if you have a ref on the kvset you don't need the tree lock, what are the use cases for these functions?

The one use case that I'm aware of for cn_tree_node_get_max_key() is for rehydrating the route map at KVS open time...